### ISSUE TITLE
The bigger the output collection the slower adding each item to history becomes

### ISSUE THREAD

**Issue by greenwoodma** [galaxyproject/galaxy#5551]
I work on the OpenMinTeD project (http://openminted.eu/) where internally we use Galaxy for running our workflows. We don't use the web UI during execution, everything is controlled by calls to the Galaxy REST API. In general this all works nicely, but we've hit an odd performance issue that seems to only occur when we use dataset collections consisting of large numbers of files (we've been testing with around 500 files and seeing this ever time).

We've attempted to narrow down the problem and have reduced it to the following example. We have a simple workflow containing two components. There is no input to the workflow as such, in that the first component has a parameter that tells it where to retrieve the input corpus from (this was done so as not to have to upload large numbers of files into a Galaxy history via the REST API). The output from the first component is one or more PDF files. These PDF files are then used as input to the second component, which in this example simply converts the PDF files into XML files (containing UIMA CAS documents), which then become the output from the workflow.

If we run the workflow accessing a corpus containing just a few documents then everything works as expected and at a sensible speed. When the first step completes looking at the Galaxy logs I can see messages such as 

`Feb 16 13:28:44 snf-1289 run.sh[5685]: galaxy.tools.parameters.output_collect DEBUG 2018-02-16 13:28:44,199 (4613) Created dynamic collection dataset for path [/srv/galaxy/database/jobs_directory/004/4613/working/out/dataset_40100.pdf] with element identifier [dataset_40100.pdf] for output [output] (40.190 ms)`

`Feb 16 13:28:44 snf-1289 run.sh[5685]: galaxy.tools.parameters.output_collect DEBUG 2018-02-16 13:28:44,239 (4613) Created dynamic collection dataset for path [/srv/galaxy/database/jobs_directory/004/4613/working/out/dataset_40101.pdf] with element identifier [dataset_40101.pdf] for output [output] (40.098 ms)`

As you can see the two PDF files are added to the history in around 40ms each. The second component then runs and when completed the following appears in the log

`Feb 16 13:29:24 snf-1289 run.sh[5685]: galaxy.tools.parameters.output_collect DEBUG 2018-02-16 13:29:24,728 (4614) Created dynamic collection dataset for path [/srv/galaxy/database/jobs_directory/004/4614/working/out/dataset_40100.pdf.xmi] with element identifier [dataset_40100.pdf.xmi] for output [output] (50.491 ms)`

`Feb 16 13:29:24 snf-1289 run.sh[5685]: galaxy.tools.parameters.output_collect DEBUG 2018-02-16 13:29:24,778 (4614) Created dynamic collection dataset for path [/srv/galaxy/database/jobs_directory/004/4614/working/out/dataset_40101.pdf.xmi] with element identifier [dataset_40101.pdf.xmi] for output [output] (49.683 ms)`

Again the XMI files produced are added to the history, this time in roughly 50ms each. These times seem perfectly acceptable and we expected the times to stay relatively stable as we increased the corpus size (i.e. adding another document of a similar size would see it added to the histories in a similar length of time). Unfortunately this isn't the case.

If instead of processing a corpus of two documents we process a corpus of 500 documents we see very different behaviour, even for the same documents. The first component runs exactly as before with the documents being added to the history in around 40ms. The second component runs, but when it completes instead of each document being added to the history in around 50ms, it takes closer to 10 seconds per document; there is quite some variability in the time but the quickest I've seen is around 4 seconds and the longest almost 20s. Looking at the log we can see this clearly for the very same documents as we saw before

`Feb 16 13:17:12 snf-1289 run.sh[5031]: galaxy.tools.parameters.output_collect DEBUG 2018-02-16 13:17:12,186 (4610) Created dynamic collection dataset for path [/srv/galaxy/database/jobs_directory/004/4610/working/out/dataset_40100.pdf.xmi] with element identifier [dataset_40100.pdf.xmi] for output [output] (5834.336 ms)`

`Feb 16 13:17:27 snf-1289 run.sh[5031]: galaxy.tools.parameters.output_collect DEBUG 2018-02-16 13:17:27,670 (4610) Created dynamic collection dataset for path [/srv/galaxy/database/jobs_directory/004/4610/working/out/dataset_40101.pdf.xmi] with element identifier [dataset_40101.pdf.xmi] for output [output] (15483.410 ms)`

Given that these are the same documents I can only imagine that the difference in time required to add them to the history is a result of the increase in the number of files to process. Now obviously I would expect the length of time to process 500 documents to be longer than that to process 2 documents, but I would assume a linear increase (i.e. still 50ms per document) instead each document takes two orders of magnitude longer to appear in the history. That means that although the processing of the 500 documents completes within a couple of minutes, we have to wait over an hour for the files to appear in the history and the workflow to finish (even if we assume an average of 10s per document that's 83 minutes just to add the documents to the history compared with less than a minute at 50ms a document).

At this point we've attempted to ensure that the problem does not lie in any of our surrounding code/infrastructure (for example all the files are local to the machine once the component finishes so it's not a network issue) and we are now at a loss as to what could be causing this huge decrease in performance. Any suggestions would be most appreciated.

**Comment by mvdbeek** [comment]
Hi @greenwoodma, nice to see that this (mostly) works out for you. Which version of galaxy are you running?
The tool you're using is dynamically producing outputs, this is the hardest kind of collection to create in galaxy. One of the things that happens at this step: `Created dynamic collection dataset for path ` is copying the created element from a temporary location to its final location, and this is done by the job handler. So what could happen here is many simultaneous copy processes getting in each others' way. Are you using the regular disk-based object store ?

In that case can you try [this patch](https://gist.github.com/anonymous/039ef8c6f6560a51f32ba8a950871470) and see if things are more performant?
Alternatively (and preferably) if you can predict how many outputs you're going to produce you may be able to pre-determine the output structure, in which case your cluster nodes can do the moving/copying.

**Comment by greenwoodma** [comment]
Thanks @mvdbeek for the quick reply. I can't believe I managed to miss off the version we are using, sorry. We are currently using 17.09. We did try updating to a newer version but had some API issues, so we rolled back for now to the version we had been using for a while.

I tried the changes in the patch (had to apply them manually due to the differing versions etc.) and they didn't seem to make any difference, but thanks for the suggestion.

What I'm most confused by is the fact that both components in the workflow produce the same number of files, but the first component runs quickly (around 40ms per file to add to the history) with the second one always taking a lot longer (10s per file or thereabouts) on the larger datasets. Interestingly the first component is run locally on the machine running Galaxy whereas the second is run on a cluster, but by the time it tries to add the files to the history they are, in both cases, present in the output folder on the machine running Galaxy, so I don't think the problem is related to where the tool runs.

As we don't know the size of a corpus in advance we use the dynamic collections to find all the files. If you think this is the issue, then I'm sure we could add something to our tools to produce a list of files to add to the history, or move them to a specific location etc.. I assume this is what you meant by "pre-determine the output structure". Is there any documentation on how we might do that?

**Comment by mvdbeek** [comment]
> I tried the changes in the patch (had to apply them manually due to the differing versions etc.) and they didn't seem to make any difference, but thanks for the suggestion.

Alright, I thought this might perhaps help directly, but this patch would only make a difference if a move is faster than a copy, which might not be a case if the move effectively copies and deletes the file (i.e when crossing file systems), or if we're saturating IO with moves as well.

The timing you were referring to is being taken between these lines:
https://github.com/galaxyproject/galaxy/blob/da002bdf44ff9c3809c15990e5f086d029c34533/lib/galaxy/tools/parameters/output_collect.py#L225-L256

The only heavy operation I could spot there is the `create_dataset ` call, which copies (or moves with the patch) the individual discovered output datasets.
I'd still be curious to know if the dataset_creation is the slow part, could you drop in a log statement around this line:
https://github.com/galaxyproject/galaxy/blob/da002bdf44ff9c3809c15990e5f086d029c34533/lib/galaxy/tools/parameters/output_collect.py#L313 ?

Something like:
```python
        timer = ExecutionTimer()
        app.object_store.update_from_file(primary_data.dataset, file_name=filename, create=True)
        log.debug("Updated object store with file '%s'. %s" , filename, timer)
```

My alternative suggestion would also only help if this is indeed the bottleneck.

If that is the case you could have a tool that dynamically creates a dataset list, where each item in the list is very small, and so the overhead of running the file creation code is small. Then the second tool can consume one item at a time, producing a constant amount of outputs per input. The advantage here is that galaxy then can before submitting the job provide the correct number of object store items directly to the job, which can write to the final output location, so no extra moving on the galaxy head node would be involved after completing the job on the cluster.

**Comment by mvdbeek** [comment]
Oh, and if you encounter an issue with the API in 18.01 (which should be stable across versions) we're happy top take a look at that as well.

**Comment by greenwoodma** [comment]
Thanks, I'll dig into this in more detail after the weekend, but now I know where I'm looking (doesn't help that python isn't my strong suit) I'll get some extra debug messages added so we can confirm exactly where the time is being taken.

As to the API issues with a later version, I'm not sure exactly what they were (I wasn't directly involved in trying a newer version) but yes, it's on my list of things to look at, and any bugs we find I'll report them as separate issues (I think we've previously reported a couple of API things which got fixed quickly, thanks!).

**Comment by jmchilton** [comment]
Thanks for initial places to look @mvdbeek - I'll wait and see how this plays out. I just had three random comments:

- The non-linear growth of the these things makes me think the problem is not with the file copies but with keeping too many database bound objects in memory or me doing something quadratic that doesn't need to be.
- > These PDF files are then used as input to the second component, which in this example simply converts the PDF files into XML files (containing UIMA CAS documents), which then become the output from the workflow.

  If there is a one-to-one correspondence like this - using dynamic collection discovery is rarely the right choice IMO. It sounds like you should have a tool that takes in a single output and produces a single output and let Galaxy map over the tool. There is some overhead with that and you could use "structured_like" instead to run this all in one job. I'd review [this section](http://planemo.readthedocs.io/en/latest/writing_advanced.html#creating-collections) of the Planemo docs.

- I'd love to hear more about the 18.01 API breakages - it should work in backward compatible way I think. @mvdbeek has caught some things in recent weeks that I did break. At any rate - the performance profile of these operations may be considerably different in 18.01 given the huge changes to collection in that release - it might be more worthwhile trying to optimize things in that branch or may already have a fix. That said - if the problem is dynamic collection discovery - that is kind of different than the rest of the collection handling and so it may have its own unexplored quirks. I would not be surprised if there is some serious database optimization we could do for that operation.

**Comment by greenwoodma** [comment]
Thanks for the comments @jmchilton. I must admit that I can't see it being as straightforward as a file copy problem either. Having tried to move files around from the terminal I can't see any issues so I cant see why Galaxy would struggle either.

I must admit our original Galaxy integration did indeed take the approach of having the tools take one file in and produce one file as output. We switched to using collections on the assumption that in our situation it would be faster, as our tools are packaged as docker images and run on a cluster, and the overhead of reserving a node and starting the docker image could be quite significant, compared to the run time for processing a single file. Using collections means we can start the docker image once and then process the whole collection. It maybe we have to revisit this assumption.

What I still don't understand though is why the first step of the workflow behaves so differently to the second. Both produce the same number of files yet each file is added to the collection in around 40ms after the first component, but 10 seconds or so after the second. There must be something different between these two cases. As I said the first runs locally on the Galaxy server whereas the second runs on the cluster, but I can't see why that should be an issue as the files end up local to the Galaxy instance before the collection is built (as far as I can tell at least).

The issues we had with 18.01 are probably us doing something silly. I wasn't directly involved in trying 18.01 but my understanding was there was some issue when they tried to run a workflow from our infrastructure, and because we had an meeting and hackathon scheduled the system was kept on 17.09 rather than investigated further (not sure even what the actual issue they hit was). It may be that we have to move to 18.01 to fix this issue with collections, but even if not I'll have another look at this at some point, but it's probably something odd we are doing rather than a regression on your side.

**Comment by mvdbeek** [comment]
I agree it seems strange that copying too many files would lead to an exponential slowdown, but I have seen this happen before when copying files over the network. If the timing says otherwise I'm happy to stop insisting.

> as the files end up local to the Galaxy instance before the collection is built (as far as I can tell at least).

For dynamic collections these files should be in the job working directory, so there's still going to be some copying involved.

**Comment by greenwoodma** [comment]
Yes, the files are in the job working directory, and I can copy or move them from there quickly from a shell. Out of interest where would Galaxy try and copy them to so I can check the destination folder isn't acting oddly.

Hopefully I'll have some numbers to report tomorrow from the logging (it's a public holiday in Greece where my colleague who's working on this with me is based)

**Comment by mvdbeek** [comment]
> Out of interest where would Galaxy try and copy them to so I can check the destination folder isn't acting oddly.

code-wise that happens where I suggested to add the logging above, in /lib/galaxy/tools/parameters/output_collect.py,  `app.object_store.update_from_file(primary_data.dataset, file_name=filename, create=True)`

The output location is being created in that call as well and depends on the object store.

**Comment by greenwoodma** [comment]
Perfect, thanks.

**Comment by greenwoodma** [comment]
Okay, so after lots of digging around and a lot of new timers and debug statements we've managed to narrow down the location of the problem. It is inside `output_collect.py` but strangely it doesn't seem to be the call to `update_from_file` but that causes the problem but the call to `_new_hda`. We put a timer around the call and can see that it's what is taking the time.

What's really weird though is that we then put timers around the calls inside that function and none of them explain the slow performance. The only thing left is the `return primary_data` call. I'm no python expert but I can't see why this would take so long, unless internally it's having to copy lots of data in memory.

I did find [one similar issue being discussed on stackoverflow](https://stackoverflow.com/questions/43529023/function-takes-a-long-time-to-return) but not sure if it's exactly related.

Currently we are trying to inline the function to avoid the return to see if that improves performance. Will report back when I have some more details.

**Comment by mvdbeek** [comment]
I find the profilehooks.profile decorator very helpful for looking at slow calls, I wrote a small piece on that [here](https://github.com/galaxyproject/galaxy/blob/dev/doc/source/dev/finding_and_improving_slow_code.rst), maybe you could decorate the call to `create_dataset` ? That should tell you exactly where how much time is being spent.

**Comment by greenwoodma** [comment]
Thanks, I'll have a look at trying that at some point to help debug the underlying issue. We have, however, now isolated the problem and, at least temporarily, been able to work around it. This message might be quite long as there will be code and log messages but hopefully it will help someone figure out what's going on.

So firstly, the return from `_new_hda` seems to have been a bit of a red herring. When we inlined that function we still had the same speed issues. So we added more log messages, pretty much a timer and debug message either side of every line of code.

```python
        timer = ExecutionTimer()
        # AF replacing following line
        #primary_data = _new_hda(app, sa_session, ext, designation, visible, dbkey, self.permissions)
        # with the following modified C&P from the function
        timer1 = ExecutionTimer()
        primary_data = app.model.HistoryDatasetAssociation(extension=ext,
                                                           designation=designation,
                                                           visible=visible,
                                                           dbkey=dbkey,
                                                           create_dataset=True,
                                                           flush=False,
                                                           sa_session=sa_session)
        log.debug("omtd: primary data '%s'. %s" , primary_data, timer1)
        timer3 = ExecutionTimer()
        if self.permissions is not UNSET:
            app.security_agent.set_all_dataset_permissions(primary_data.dataset, self.permissions, new=True, flush=False)
        log.debug("omtd: dataset permissions '%s'. %s", app.security_agent, timer3)
        timer2 = ExecutionTimer()
        sa_session.add(primary_data)
        log.debug("omtd: security agent '%s'. %s" , app.security_agent, timer2)
        # end modified C&P

        log.debug("omtd: new hda '%s'. %s" , sa_session, timer)
```

which resulted in timings in the logs such as this set

```
Feb 20 14:05:11 snf-1289 run.sh[9011]: galaxy.tools.parameters.output_collect DEBUG 2018-02-20 14:05:11,520 omtd: primary data '<galaxy.model.HistoryDatasetAssociation object at 0x7f9ae8733810>'. (0.585 ms)
Feb 20 14:05:17 snf-1289 run.sh[9011]: galaxy.tools.parameters.output_collect DEBUG 2018-02-20 14:05:17,135 omtd: dataset permissions '<galaxy.security.GalaxyRBACAgent instance at 0x7f9bdc3c2e18>'. (5613.836 ms)
Feb 20 14:05:17 snf-1289 run.sh[9011]: galaxy.tools.parameters.output_collect DEBUG 2018-02-20 14:05:17,135 omtd: security agent '<galaxy.security.GalaxyRBACAgent instance at 0x7f9bdc3c2e18>'. (0.276 ms)
Feb 20 14:05:17 snf-1289 run.sh[9011]: galaxy.tools.parameters.output_collect DEBUG 2018-02-20 14:05:17,136 omtd: new hda '<sqlalchemy.orm.scoping.scoped_session object at 0x7f9bdc429550>'. (5615.783 ms)
```

Not being a python person, and not having a good understanding of the internals of Galaxy I can't from the code understand why setting the permissions should take so long. And why it only does take so long when processing the output of the second step not the first step. Here is the same log messages for one document from the first step

```
Feb 20 13:54:55 snf-1289 run.sh[9011]: galaxy.tools.parameters.output_collect DEBUG 2018-02-20 13:54:55,482 omtd: primary data '<galaxy.model.HistoryDatasetAssociation object at 0x7f9ac879f2d0>'. (0.749 ms)
Feb 20 13:54:55 snf-1289 run.sh[9011]: galaxy.tools.parameters.output_collect DEBUG 2018-02-20 13:54:55,487 omtd: dataset permissions '<galaxy.security.GalaxyRBACAgent instance at 0x7f9bdc3c2e18>'. (4.972 ms)
Feb 20 13:54:55 snf-1289 run.sh[9011]: galaxy.tools.parameters.output_collect DEBUG 2018-02-20 13:54:55,488 omtd: security agent '<galaxy.security.GalaxyRBACAgent instance at 0x7f9bdc3c2e18>'. (0.285 ms)
Feb 20 13:54:55 snf-1289 run.sh[9011]: galaxy.tools.parameters.output_collect DEBUG 2018-02-20 13:54:55,488 omtd: new hda '<sqlalchemy.orm.scoping.scoped_session object at 0x7f9bdc429550>'. (6.249 ms)
```

Most of the time is still inside the permissions code but it's 4.972 ms not 5613.836 ms! Fortunately because of the way we use Galaxy it would seem we can simply comment out the if statement and the outputs are still correctly loaded into the history and made available by the REST service, so that's a workaround we can adopt for now.

I don't have any real idea why setting the permissions should be so slow. All I can think is that for some reason each time that is called, instead of setting the permissions on the single file being processed it's doing so over every file in the collection. In otherwords the roughly 5000ms is actually made up of 10ms for each of the 500 documents. The maths seems to roughly add up, but I can't work out from the code if this is likely or not. This would also explain why it works with a collection of just two documents, as we wouldn't notice the extra 10ms added to the time to handle each file.

I hope that is all helpful information, and we will keep digging as it would be nice to know exactly what's going on. So if you need any more information or for us to run any tests etc. please do let us know and we'd be happy to help. Thanks again @mvdbeek  and @jmchilton for all the help and comments, most appreciated.

**Comment by mvdbeek** [comment]
I think that is very helpful, thanks a lot @greenwoodma! I had a quick look at this and it seems we're getting the same set of permissions over-and-over in the loop over the discovered output dataset, which we should cache. There are a few targets in there that should be optimised, though it's not quite apparent why this would become successively slower.

**Comment by greenwoodma** [comment]
I suppose it doesn't really get successively slower in that the first document in a 500 document history takes roughly the same amount of time to process as the last document, but both take a lot longer than the first and last in a history containing just two documents. I guess if you are resetting and then applying the permissions for all 500 entries each time around then that would certainly explain things.

I wonder if rather than caching the data, the code should simply be moved outside of the loop and done once after all the entries in the history have been processed?

**Comment by mvdbeek** [comment]
Probably, and we could also create the datasets in a single transaction, since we by then know how many datasets to create.

**Comment by jmchilton** [comment]
I'll take a look at these optimization ideas today - thanks @greenwoodma and @mvdbeek !

**Comment by jmchilton** [comment]
> we could also create the datasets in a single transaction

Sounds possible but we would need to refactor this code even more than that I think, we would need to pre-create all the datasets and then run all the update_from_file commands - since I think we need to have dataset IDs before we can do update_from_file.

Update: ...and I think why I didn't do that last time I was trying to optimize this is that I thought update_from_file would dominate the time spent in this method anyway.

**Comment by mvdbeek** [comment]
`update_from_file` being most expensive was my instinct as well, but I haven't even thought of discovering collection outputs as a way to "batch" small runs together, that is definitely a cool idea.
Did you add some measurements in #5578 to see how that would affect large discovered outputs ?

**Comment by mvdbeek** [comment]
@greenwoodma I think https://github.com/galaxyproject/galaxy/pull/5578 should address the permissions issue for 18.01. I understand you're not yet migrating, but please let us know once you do if this has resolved the issue.

**Comment by greenwoodma** [comment]
@mvdbeek yes, currently we are staying on 17.09 for stability reasons while we do some further testing work. I will try running a copy of 18.01 locally at some point with the fix in place to see if I can confirm it works for us. If not then we will probably move to 18.01 at some point anyway at which point we'll revist the issue.

Thanks again for look into this so quickly and to @jmchilton for patching it so quickly.

**Comment by jmchilton** [comment]
Please reopen or create a new issue if this becomes a bottleneck again, I think there are still bottlenecks but hopefully they have shifted around and we have issues open to address them.
